flang/lib/Optimizer/Transforms/PolymorphicOpConversion.cpp

```diff
@@ -229,7 +229,9 @@ class PolymorphicOpConversion
 
   void runOnOperation() override {
     auto *context = &getContext();
-    auto mod = getOperation()->getParentOfType<ModuleOp>();
+    auto mod = mlir::dyn_cast_or_null<mlir::ModuleOp>(getOperation());
```

Comment:
I believe we can remove the mutex, just a few lines above, if it is to be a module pass.

Reply:
It is run on a module when invoked with fir-opt in tests, but in the real pipeline it is run in parallel on each top level operation (see the tests printing the pass pipeline)

Reply:
My plan was to make this pass a module pass and remove the mutex. The reason is that if there are going to be multiple instances of this pass running in parallel on, say, all functions, then the mutex will be effective at preventing races since each instance maintain its own. You would need a shared mutex for that. But a global mutex is not something MLIR allows:
https://mlir.llvm.org/docs/PassManagement/#operation-pass
```
Must not maintain any global mutable state, e.g. static variables within the source file. All mutable state should be maintained by an instance of the pass.
```

So, that was my plan.

Reply:
I agree with Renaud that the mutex is not functional as-is.  Each instance of the pass will have its own mutex, so the parallel modifications of the parent module operation cause a race.

Reply:
To be clear, the pass already ran in parallel over all functions, this change just added a few other top level operations.

I will change it to a module pass.

Reply:
https://github.com/llvm/llvm-project/pull/90721